File: MergeSort.cs
################################################################################
#
# Copyright (c) 2018
# Authors:
#  Hesham M. Eraqi (heraqi@aucegypt.edu)
#
################################################################################

using System;
using System.Collections.Generic; // For Lists and Dictionaries

// To execute C#, please define "static void Main" on a class
// named Solution.

class Node {
    public int key;
    public string data;
    
    public Node(int k, string d){
        key = k;
        data = d;
    }
    
    public Node(Node anotherNode){
        key = anotherNode.key;
        data = anotherNode.data;
    }
}

class Solution
{
    static void Merge(List<Node> A, int s, int m, int e) {
        List<Node> B = new List<Node>(A); // Copy A to B
        List<Node> sorted_part = new List<Node>();

        List<Node> L1 = B.GetRange(s,m-s+1);
        List<Node> L2 = B.GetRange(m+1,e-(m+1)+1);
        while (L1.Count >0 && L2.Count > 0) {
            if (L1[0].key <= L2[0].key) {
                sorted_part.Add(new Node(L1[0]));
                L1.RemoveAt(0);
            }
            else {
                sorted_part.Add(new Node(L2[0]));
                L2.RemoveAt(0);
            }
        }
        
        foreach (Node n in L1)
            sorted_part.Add(n);
        foreach (Node n in L2)
            sorted_part.Add(n);
        
        // Add the sorted_part to A
        for (int i=0; i<e-s+1; i++)
            A[i+s] = sorted_part[i];
    }
    
    static void MergeSort(List<Node> A, int s, int e) {
        if (s < e) {
            int m = (e+s)/2;
            MergeSort(A, s, m);
            MergeSort(A, m+1, e);
            Merge(A, s, m, e);
        }
    }
    
    static void Main(string[] args)
    {
        List<Node> data = new List<Node> {
            new Node(10, "Hesham"),
            new Node(5, "Ahmed"),
            new Node(7, "John"),
            new Node(8, "Eman"),
            new Node(10, "Nikki"),
            new Node(20, "Sarah"),
            new Node(9, "Mai"),
            new Node(8, "Michael"),
            new Node(4, "Fred"),
            new Node(3, "Jens")
        };
        
        // O(nlog(n)) - Stable - Not Inplace
        MergeSort(data, 0, data.Count-1);
             
        foreach (Node n in data)
            Console.Write(n.key.ToString() + "_" + n.data + " ");
    }
}

File: MergeSort.py
################################################################################
#
# Copyright (c) 2020
# Authors:
#  Hesham M. Eraqi (heraqi@aucegypt.edu)
#
# Mergesort: O(nlog(n)) - Stable (similar keys retains their initial order) - Not Inplace
# Quicksort: O(n^2) worst (O(nlog(n)) best&average) - Not Stable - Inplace
#
################################################################################
class Node:
    def __init__(self, key, data):
        self.key = key
        self.data = data

def merge_sort(data, s, e):
    if e > s: # Not a single element
        m = s + int((e - s)/2)
        merge_sort(data, s, m) # First part might be larger by 1
        merge_sort(data, m+1, e)
        merge(data, s, m, e)
        
def merge(data, s, m, e):
    data_copy = data.copy()
    L1 = data_copy[s:m+1]
    L2 = data_copy[m+1:e+1]
    sorted_part = []
    while (len(L1) >0 and len(L2) > 0):
        if (L1[0].key <= L2[0].key):
            sorted_part.append(Node(L1[0].key, L1[0].data))
            L1.pop(0)
        else:
            sorted_part.append(Node(L2[0].key, L2[0].data))
            L2.pop(0)
    for n in L1:
        sorted_part.append(n)
    for n in L2:
        sorted_part.append(n)

    # Add the sorted_part to A
    for i in range(len(sorted_part)):
        data[i+s] = sorted_part[i]
        
data = [Node(10, "Hesham"),
        Node(5, "Ahmed"),
        Node(7, "John"),
        Node(8, "Eman"),
        Node(10, "Nikki"),
        Node(20, "Sarah"),
        Node(9, "Mai"),
        Node(8, "Michael"),
        Node(4, "Fred"),
        Node(3, "Jens")]
merge_sort(data, 0, len(data)-1)  # takes start and end indices
print([str(data[i].key) + " " + data[i].data for i in range(len(data))])
